Add Pin Lock Timeout Option

roundhill opened the pull request:
Adding a new setting for the pin lock timeout that will dismiss the lock screen if the chosen amount of time hasn't passed yet. I'm not particularly fond of how much code it took to add this feature, but it does appear to work nicely :)

Screenshot:
![simulator screen shot - iphone 7 - 2018-08-22 at 13 57 35](https://user-images.githubusercontent.com/789137/44490743-fb916780-a613-11e8-9871-005f614ea73d.png)

**To Test**
* Enable the passcode lock in Settings.
* Leave the app and come back, you should immediately see the passcode lock.
* Change the `Lock Timeout` setting to something other than `Off`.
* Leave the app and come back within the timeframe that you set. You should return to the app where you left off with no pin lock presented.
* Setting it back to `Off` should imeddiately show the pin lock again when returning to the app.

I've been 🤔 about this related to security. I suppose you could change the clock of the device to try and match when the user last left the app to bypass the lock screen, but you'd have to know when they last left the app so I'm not too worried about that. Any other security considerations the reviewer may have is most welcome!

Fixes #215

jklausa commented:
How about using a monotonic clock like http://pubs.opengroup.org/onlinepubs/9699919799/functions/clock_gettime.html? It'll guard against someone changing the date/time manually _and_ more importantly, would get around bunch of weird behaviors users might see when traveling across timezones.

jklausa commented:
(Should be available on iOS since 10+.)

jklausa commented:
I'll admit I am not 100 % sure what security promises we're trying to make with the pin screen in the first place, but for full-paranoia-mode I'd think about maybe also storing the timeout preference in Keychain.

roundhill commented:
> I'd think about maybe also storing the timeout preference in Keychain

Great idea, I'll do that 👍

jleandroperez commented on Simplenote/Classes/SPOptionsViewController.m:
```diff
@@ -313,6 +350,15 @@ - (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(N
+                    
+                    NSInteger timeoutPref = [[NSUserDefaults standardUserDefaults] integerForKey:kPinTimeoutPreferencesKey];
+                    [cell.detailTextLabel setText:timeoutPickerOptions[timeoutPref]];
+                }
```
Same as above, missing `break;` on this one. (Nothing breaks though)

roundhill replied:
Because of how we hide the `biometry` row, I actually needed to leave this break in here. I added a comment in the code. I'm open to alternative ideas :)

jleandroperez commented on Simplenote/Classes/SPOptionsViewController.m:
```diff
@@ -313,6 +350,15 @@ - (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(N
                     cell.accessoryView = self.biometrySwitch;
                     cell.tag = kTagTouchID;
```
Missing break on this one `break;`.

(ObjC's behavior is to fall thru unless there's a break)

jleandroperez commented on Simplenote/Classes/SPOptionsViewController.m:
```diff
@@ -649,4 +708,33 @@ - (BOOL)themePref
+- (NSAttributedString *)pickerView:(UIPickerView *)pickerView attributedTitleForRow:(NSInteger)row forComponent:(NSInteger)component {
+    VSTheme *theme = [[VSThemeManager sharedManager] theme];
+    NSAttributedString *attributedTitle =
+    [[NSAttributedString alloc] initWithString:timeoutPickerOptions[row] attributes:@{NSForegroundColorAttributeName:[theme colorForKey:@"textColor"]}];
```
Maybe put the allocation and the assignment in the same line? (and move the `attributes` to a variable to shorten the line length?)

jleandroperez commented on Simplenote/SPAppDelegate.m:
```diff
@@ -308,6 +308,48 @@ - (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(
+    
+    return [timeoutValues[timeoutPref] integerValue];
+}
+
```
Nipticky Suggestion! how about moving `shouldBypassPinLock` + `getPinLockTimeoutSeconds` (and as much as possible, all related to the Pinlock!) to a class to contain all that logic?

`PinLockManager` || `PinLockController` || `LockController` || `LockManager` || `Anything better!`

Goal is to help reduce a bit the AppDelegate and contain all of the related code in a single spot.

roundhill replied:
Haha I thought "I bet @jleandroperez will want this code to be moved to a static class" at one point ;)

roundhill replied:
Added a new swift class for this 👍

roundhill commented:
@jleandroperez Thank you! Ready for another look 👀

roundhill commented:
> As a result, the FaceID lock will be onscreen (even though it should have been bypassed!)

Yikes! Fixing this, good catch.

roundhill commented:
@jleandroperez Ready once again :) Thanks.

roundhill commented:
ok @jleandroperez I dug deep and fixed up a few things. Mainly I should have used `applicationDidBecomeActive` vs. `applicationWillEnterForeground`. I also fixed a pre-existing bug where the touch/face Id prompt would not show during certain situations. Here's a good test list:

- Swiping up to reveal control center should show face Id when returning to the app. Same for entering the multitask view. The timeout should work here as well, dismissing the pin lock if within the timeout period.
- Starting the app cold should show the pin lock or bypass it where appropriate.
- Going back to the home screen and back to the app multiple times should work as expected.

roundhill commented:
Thank you once again @jleandroperez!
